### ISSUE TITLE
Dynamic build plan generation

### ISSUE THREAD

**Issue by vito** [concourse/concourse#684]
This proposal is to add `load: foo.yml` as a primitive to build plans, allowing parts of the build plan to be dynamically generated while the build is running.

Original discussion:
- https://github.com/concourse/concourse/issues/534#issuecomment-235650894 (and followup comments).

This is a single change that would allow for the following:
- #292 (dynamic build matrices)
  
  ``` yaml
  - task: generate-matrix
  - aggregate: {load: generated-matrix/steps.yml}
  ```
- #300 (dynamic worker placement)
  
  ``` yaml
  - put: workers
    params: {acquire: true}
  - task: some-task
    tags: [{load: workers/name}]
  ```
- #210 (dynamic version pinning)
  
  ``` yaml
  - get: valid-versions
    trigger: true
  - aggregate:
    - get: concourse
      version: {load: valid-versions/concourse.json}
    - get: garden-runc
      version: {load: valid-versions/garden-runc.json}
  ```

As well as allowing resources to be dynamically configured:
- concourse/bosh-deployment-resource#13 - dynamic deployment name
- concourse/bosh-deployment-resource#6 - dynamic target address
- concourse/cf-resource#20 - dynamic space name

There are some interesting implications to this, however.
- We can no longer know the whole build plan up-front, so we'd need to dynamically render that. Might not be too bad; maybe we just keep saving the build plan as it changes, and the UI keeps up.
- Dynamically created `get` steps obviously won't be able to trigger the job. I don't think this matters much as there'd probably always be a higher-level `get` (i.e. of the set of things to fetch) that would really trigger it anyway.
- Config validation and parsing gets tricky, now that every value can technically be replaced with `{load: ...}`.

**Comment by jchesterpivotal** [comment]
Broadly: I'm wary of this change, as it weakens the declarative WYSIWYGness of the existing model.

Also unclear: when is the master plan evaluated to render the final plan? When I use `fly`? Or on atc?

If it's on `fly`, then I can at least ask it to show me the fully evaluated plan before I commit it. It could be broken into two steps: `fly evaluate-pipeline` and `fly apply-pipeline`, which `fly set-pipeline` acting as an alias for these.

More narrowly, I'm not sure it solves the particular issues referenced in the way folks would like to solve them.

**Comment by vito** [comment]
@jchesterpivotal The plan is evaluated as the build is running. There's no other way around it since the whole mechanic of `load` is to refer to dynamically created (either `get` or a `task` output) data. Given that, I don't expect this to be used excessively; it's not meant to be syntactic sugar, it's meant to support these occasions where the plan (or, minimally, certain parameters within the plan) cannot be statically configured ahead of time.

This is already a very common thing today with resources, which often allow a static config and also loading a value from a file. This issue would remove all that duplication from resources.

For #210 at least I chatted with @cppforlife and we both agreed it was a reasonable solution. For #300 it's pretty much exactly what the issue requested. Not sure about #292 since that issue was opened a long while ago and had a bit of discussion, but the whole point of this proposal is to discuss it, so now's the time.

**Comment by amitkgupta** [comment]
Dynamic bosh deployment configuration would be sweet for pipelines that automate IaaS and BOSH provisioning! /cc @JesseTAlford

**Comment by vito** [comment]
Interestingly this would remove the need for `file:` on task configs. You could do `config: {load: ...}` instead.

We'd obviously keep it around for a long time for backwards compatibility.

**Comment by JesseTAlford** [comment]
I do like that it gets rid of `file:`; a more generalized alternative to inline configuration that handles tasks and other things alike is great, and makes that whole situation clearer. I like it when something's that special becomes not-special.

This would help us clean up our pipelines. We currently have extra complexity where we write out target files, pass them between tasks, then read them in as env vars in subsequent tasks. I would love to see that nonsense go.

**Comment by cppforlife** [comment]
cc @dsabeti

**Comment by zachgersh** [comment]
@vito how do the implicit `gets` that happen on put get this new dynamic build-plan data? One of the problems I had to address with the `director_file` PR above is that a `put` can end up using dynamic configuration that the implicit `get` then knows nothing about (resulting in a failure).

**Comment by vito** [comment]
@zachgersh You could either propagate it through the version emitted by the resource, or set it via `get_params`. Tucking it into the version would be a bad idea for secrets, but it does make the resource self-contained and in principle able to propagate through the pipeline statically. Otherwise you'll just need to keep fetching the dynamic bits and using `params` to pass things along.

Either way we're better off with this than with `*_file` params as you have the option of dynamically passing information to `get_params`.

**Comment by rfliam** [comment]
This may seem specious... but:
```- aggregate: {load: generated-matrix/steps.yml}```
Would make pipelines turning complete because you could perform recursion. Which, uh, seems undesirable.

**Comment by ljfranklin** [comment]
Big +1 for dynamic `gets`. We have a ton of bash tasks that read a value from a file and download something based on a value in that file. For example, reading the Terraform statefile to determine which IaaS stemcell we should download. Would be much nicer to do the download part with a first-class resource and keep our tasks side-effect free. I also had to hack around this in the terraform-resource by adding a `put.env_name_file` and telling everyone to always use `put` instead of `get`.

**Comment by renbeynolds** [comment]
This type of feature would be a huge plus for us. As a workaround, we've had to resort to actually re-flying the pipeline from within itself to dynamically add `put` steps to our build plans. Many of our pipelines apply configuration stored within a repo and it's a common pattern to want to add pipeline steps based on the presence of additional configuration files (say for different apps or teams that need to be configured / deployed).

**Comment by mwmahlberg** [comment]
My two cents as a user on build matrices and worker selection. The version pinning aspect eludes me a bit, so I refrain from commenting on it.

# Dynamic build matrices
As for dynamic build matrices, this can already be done with a combination of `get`s and task scripting.  In its most simple form, one could use a git resource and a matrix branch to contain an env file
   
 ```
 ARCHES=amd64 arm
 DBMS=sqlite bolt
 ```

and a task

```bash
. getoutput/env

ARCHES=${ARCHES:-"amd64"}
DBMS=${DBMS:-"bolt"}

for ARCH in $ARCHES; do
  for DB in $DBMS; do
    echo "Building $ARCH-$DB"
    # do something fancy
  done;
done;
```

So instead of adding complexity to a deliberately simple concept, the dynamic part can be achieved either by existing - or in more extreme cases custom - resource-types. This would have the added bonus that when the dynamic part changes (say an arch is added to the above example), it could be used as a trigger.

Granted, the above gets only displayed as a single task, but I actually think of that as an advantage: If there is a build matrix, the expected outcome is that either _all_ combinations succeed or _all_ fail, no? The wish to have all entries in a build matrix to be visualized independently probably stems from the idea to find the "smoking gun" by means of visualization. As per informational value, nothing would change. But, as said, these are only my thoughts.

## Dynamic worker placement

As for the dynamic worker placement I can not say too much, as I let the system deal with selecting the workers.

But maybe this is quite the point: If I had to choose the workers dynamically, then I would rethink wether my tags match my workloads rather than changing the content of `tag` dynamically.

The original use case in #300 does not convince me. It is not that the workers would be spun up dynamically. So in this case, dynamic worker placement would only be there to prevent a build on an IaaS worker _if_ there are still resources available locally. But that kind of management sounds more like a job for an underlying orchestration: check wether the workers are at their limit, dynamically add a _node_ on IaaS and subsequently a worker on said node if they are. Let the orchestration and concourse do their respective jobs.

**Comment by rhoughton-pivot** [comment]
We have a case where there is a pool of `darwin` workers, and would like to check them out  using a `pool` resource, and then dynamically set that tag on the subsequent task. How is that use case?

**Comment by vito** [comment]
@rhoughton-pivot Yeah, that's one of the common use cases I've heard (the issue description references #300). This kind of feels like something that could be accomplished through anti-affinity though, which would eliminate the need for a `pool` resource. Is the goal to make sure your Darwin workers don't run e.g. more than one Xcode build at a time?

---

As a broader update, I think the need for full-blown dynamic build plan generation is being slowly chipped away through the introduction of the `load_var` step and the broader pipeline automation initiatives planned out in v10. I think it'd be pretty easy to e.g. support interpolation for the `tags:` field without having to go full-blown dynamic build plan generation.

**Comment by rhoughton-pivot** [comment]
@vito Our `darwin` jobs use `xcode`, `renv`, and `nvm` (Node Version Manager) to ensure correct dependency and build-tool versions. Running only one job at a time is imperative until Mac gets containerization.  Can you explain more about how to set anti-affinity for selection of otherwise identical workers?

**Comment by vito** [comment]
@rhoughton-pivot The only way to do it right now is to limit it across *all* workers, which isn't the greatest. It can be done by configuring the `limit-active-tasks` container placement strategy: https://concourse-ci.org/container-placement.html#limit-active-tasks-strategy

It would be nice if that limit could be configured per-worker or something so it doesn't have to be so low for all the builds that don't need it.

**Comment by rhoughton-pivot** [comment]
Or even to have a parameter on the worker binary itself that stated the number of concurrent tasks it was willing to take...

**Comment by asavchkov** [comment]
@mwmahlberg Am I right that all dynamic tasks in your matrix  (`echo "Building $ARCH-$DB"` and so on) are executed on the same runner sequentially, that is one after another? Normally a matrix build assumes that its matrix elements are executed in parallel to reduce the build time. Can we obtain a parallel execution using your approach?

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
